src/main/scala/sangria/http/akka/SangriaAkkaHttp.scala

```diff
@@ -0,0 +1,111 @@
+package sangria.http.akka
+
+import akka.http.scaladsl.model.MediaTypes._
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.{Route, StandardRoute}
+import akka.http.scaladsl.unmarshalling.{FromRequestUnmarshaller, FromStringUnmarshaller}
+import Util.explicitlyAccepts
+import sangria.ast.Document
+import sangria.parser.{QueryParser, SyntaxError}
+import GraphQLRequestUnmarshaller._
+
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success, Try}
+
+object SangriaAkkaHttp {
+  final case class MalformedRequest(private val message: String = "",
+                                   private val cause: Throwable = None.orNull)
+    extends Exception(message, cause)
+
+  case class Location(line: Int, column: Int)
+  case class GraphQLError(massage: String, locations: Option[List[Location]] = None)
+  case class GraphQLErrorResponse(errors: List[GraphQLError])
+
+  def graphQLPlayground: Route = get {
+    explicitlyAccepts(`text/html`) {
+      getFromResource("assets/playground.html")
+    }
+  }
+
+  def formatError(error: Throwable): GraphQLErrorResponse = error match {
+    case syntaxError: SyntaxError =>
+      GraphQLErrorResponse(
+        GraphQLError(
+          syntaxError.getMessage,
+          Some(Location(syntaxError.originalError.position.line, syntaxError.originalError.position.column) :: Nil)
+        ) :: Nil
+      )
+    case NonFatal(e) =>
+      GraphQLErrorResponse(GraphQLError(e.getMessage) :: Nil)
+    case e =>
+      throw e
+  }
+
+  def prepareQuery(maybeQuery: Option[String]): Try[Document] = maybeQuery match {
+    case Some(q) => QueryParser.parse(q)
+    case None => Left(MalformedRequest(
+      """Could not extract `query` from request.
+        | Please confirm you have included a valid GraphQL
+        |  query either as a QueryString parameter, or in the body of your request.""".stripMargin)).toTry
+  }
+
+  private def prepareGraphQLPost[T](inner: Try[GraphQLRequest[T]] => StandardRoute)
+                                (implicit reqUm: FromRequestUnmarshaller[GraphQLHttpRequest[T]],
+                                 varUm: FromStringUnmarshaller[T],
+                                 v: Variables[T]): Route =
+    parameters(Symbol("query").?, Symbol("operationName").?, Symbol("variables").as[T].?) { (queryParam, operationNameParam, variablesParam) =>
```

Comment:
I'm not sure how `Symbol("variables").as[T].?` will behave. In the current implementation, we read it in as a string, TRY to parse it, and continue along with an option type (somewhat regardless of failure).

I am not sure which is better, because I have not written test cases. Ideally, if someone provides variables in the `URL` AND `body`, that's a malformed request. So potentially this is fine, and what I have currently would be TOO resilient (allowing for malformed variables in the  QueryString, falling back to the body, and finally to the `empty` value).

Really this situation of "where do I get my variables?" is only complicated by `application/graphql` support. Otherwise, I think it's safe enough to say "Either it' a GET (everything is in the QueryString) or it's a POST (everything we care about should be in the body)"

I'm happy to pick one and go with it, and simply document the behavior.

Reply:
That's a good question. It's been a long time since I really worked with akka-http, but I would guess it rejects if it fails unmarshalling.

I agree with what you wrote. 
I see that [the recommendation on implementation](https://graphql.org/learn/serving-over-http/#post-request) is a bit vague, as you pointed out.

Perhaps we can make it be three distinct cases:
1. GET; all from QS
2. POST `application/json`; all from body (I would even not take query from QS)
3. POST `application/graphql`; `variables` and `operationName` from QS.

On parsing, if variables are passed and are malformed, I think it's better to fail with a parsing error rather than excepting the request and running it with empty variables. The request will most likely fail anyway, as most likely the gql query references vars. 
It's imo much more helpful to get an error "Something is wrong with your vars formatting, can't parse" than it is to get "unbound variable". Much easier to solve the issue for the user.

I think our job here is to provide the reference, "default" implementation. Users can always implement any special edge case they need on their own. 

It should be zero effort to bootstrap a Scala GraphQL project, and it should behave in an expected way (e.g. work ootb with all the popular tooling in the ecosystem).

Reply:
I think that's a good plan forward. Since there is no clear spec, we'll go ahead and say "these are sensible defaults"